### PR TITLE
Rewrite signat with ocflib

### PR DESCRIPTION
**Pull request by matthew-mcallister**
Waiting on a pull request in ocflib and needs python-tabulate to be installed, but it works nicely in a virtualenv.

### CODE REVIEW

File: staff/acct/signat [hunk 1]
```diff
@@ -1,104 +1,173 @@
...
-    else:
-        raise ValueError(("Account not found in LDAP: " + username))
+
+from tabulate import tabulate
+
+import ocflib.account.search as search
+import ocflib.constants as constants
+import ocflib.infra.ldap as ldap
+import ocflib.ucb.directory as directory
+from ocflib.ucb.groups import *
```
Comment (chriskuehl) [inline staff/acct/signat]: wildcard imports are yucky, can you replace this?

File: staff/acct/signat [hunk 2]
```diff
@@ -1,104 +1,173 @@
...
-        raise ValueError(("Invalid DN: " + ldap_basedn))
-    if entry:
-        try:
-            return int(entry[0][1][ldap_attribute][0])
-        except KeyError:
-            raise KeyError(("Account does not have a CalNet UID: " + username))
-    else:
-        raise ValueError(("Account not found in LDAP: " + username))
+
+from tabulate import tabulate
```
Comment (chriskuehl) [inline staff/acct/signat]: nit: generally python imports are organized into three blocks, in this order:
1. stdlib imports
2. 3rd-party library imports
3. modules from this library/application
(with a blank line in between each block of imports)

so tabulate would be grouped alphabetically in the same block as the ocflib imports

but I wouldn't worry about it for now. It's a pain to do manually and pre-commit will do it for us once I get around to patching it for our use case.

File: staff/acct/signat [hunk 3]
```diff
@@ -1,104 +1,173 @@
...
+
+    for (oid, attrs) in groups.items():
+        signatories = signatories_for_group(oid)
+        if signatories:
+            found = True
+            show_signatories_for_group(oid)
+            print()
+
+    if not found:
+        print(red("No groups had signatories."))
```
Comment (chriskuehl) [inline staff/acct/signat]: I think you can probably rewrite this as:

```python
for oid, attrs in groups.items():
    signatories = signatories_for_group(oid)
    if signatories:
        show_signatories_for_group(oid)
        break
else:
    print(red("No groups had signatories."))
```

Also, maybe pass the signatories in to that function so we don't look them up twice?

File: staff/acct/signat (hunk 1 above)
Comment (matthew-mcallister) [inline staff/acct/signat]: Yeah, can't rely on pre-commit to fix up my imports on this repo.

File: staff/acct/signat [hunk 4]
```diff
@@ -1,104 +1,173 @@
...
+
+    print("Searching for signatories...")
+
+    found = False
+
+    for uid in uids:
+        groups = groups_by_student_signat(uid)
+        if groups:
+            found = True
+            show_groups_by_student_signat(uid, groups)
```
Comment (chriskuehl) [inline staff/acct/signat]: second arg should be a keyword arg?

File: staff/acct/signat [hunk 5]
```diff
@@ -1,104 +1,173 @@
...
+            print()
+
+    if not found:
+        print(red("No signatories found."))
```
Comment (chriskuehl) [inline staff/acct/signat]: you can rewrite the for loop similar to above if you care to

File: staff/acct/signat (hunk 2 above)
Comment (matthew-mcallister) [inline staff/acct/signat]: I was under the impression things got organized like:
1. stdlib
2. 3rd-party external to the developer
3. other libs from the same team/codehouse
4. this codebase

The convention you suggest makes enough sense though.

**Comment by chriskuehl** [comment]
lgtm, you'll want to use apt-dater to ensure ocflib gets updated first, and you'll want to add python3-tabulate to puppet

note that python3-tabulate is not available on wheezy so for the time being you'll need to put it in an if block. i think it's fine if signat doesn't work on wheezy.

**Comment by chriskuehl** [comment]
this is a pretty great change, think you can talk about this at the next BoD?

File: staff/acct/signat (hunk 3 above)
Comment (matthew-mcallister) [inline staff/acct/signat]: I don't think that will result in the same behavior. Won't that terminate when the first group with signatories is found?

And the latter is just a typo, thanks for catching.

File: staff/acct/signat (hunk 2 above)
Comment (chriskuehl) [inline staff/acct/signat]: I'm not familiar with those blocks, but in any case, I probably wouldn't worry about it personally. pre-commit will fix it to some standard style when I eventually get it working here.

(The reason we can't use pre-commit in utils yet is that it currently targets files based on extension, and almost nothing in utils has an extension. But I'm [working on a branch](https://github.com/pre-commit/pre-commit/compare/master...chriskuehl:hooks-by-type) that fixes that which I'll be deploying soon to OCF as a testing ground.)

File: staff/acct/signat (hunk 3 above)
Comment (chriskuehl) [inline staff/acct/signat]: you're right, I misread your code. looks fine as-is in that case.

File: staff/acct/signat (hunk 4 above)
Comment (matthew-mcallister) [inline staff/acct/signat]: Sure, why not.

**Comment by matthew-mcallister** [comment]
I'll go over this at BoD. `approve` is basically a two-step process now.

**Comment by matthew-mcallister** [comment]
Amended and ready to merge once Puppet is updated for deployment. In the meantime, open for comments on interface, documentation, speed, etc.

**Pull request merged**